libs/timebar/src/chartsNew/common/hooks.ts

```diff
@@ -1,23 +1,85 @@
-import { useContext, useMemo } from 'react'
-import TimelineContext from '../../timelineContext'
+import { useContext, useMemo, useState } from 'react'
+import { useDebouncedCallback } from 'use-debounce'
+import TimelineContext, { TimelineScale } from '../../timelineContext'
+import { TimebarChartChunk } from '..'
 import { TimebarChartData } from './types'
 
+const filterData = (data: TimebarChartData, start: string, end: string) => {
+  return data.map((item) => {
+    const filteredChunks = item.chunks.filter((chunk) => {
+      const chunkStart = chunk.start
+      const chunkEnd = chunk.end || chunk.start
+      return chunkEnd > +new Date(start) && chunkStart < +new Date(end)
+    })
+    return {
+      ...item,
+      chunks: filteredChunks,
+    }
+  })
+}
+
 export const useFilteredChartData = (data: TimebarChartData) => {
   const { outerStart, outerEnd } = useContext(TimelineContext)
-  // TODO overflow + debounce?
-  return useMemo(() => {
-    return data.map((item) => {
-      const filteredChunks = item.chunks.filter((chunk) => {
-        const chunkStart = chunk.start
-        const chunkEnd = chunk.end || chunk.start
-        return chunkEnd > +new Date(outerStart) && chunkStart < +new Date(outerEnd)
-      })
-      return {
-        ...item,
-        chunks: filteredChunks,
-      }
-    })
-  }, [data, outerStart, outerEnd])
+
+  const [filteredData, setFilteredData] = useState<TimebarChartData>([])
+  const debouncedSetFilteredData = useDebouncedCallback(
+    (data, outerStart, outerEnd) => {
+      setFilteredData(filterData(data, outerStart, outerEnd))
+    },
+    100,
+    { maxWait: 1000, leading: true }
+  )
+
+  useMemo(() => {
+    debouncedSetFilteredData(data, outerStart, outerEnd)
+  }, [outerStart, outerEnd, data, debouncedSetFilteredData])
+
+  return filteredData
 }
 
-export const useClusteredChartData = (data: TimebarChartData) => {}
+const MIN_DISTANCE_PX_TO_CLUSTER = 2
+
+const clusterData = (data: TimebarChartData, outerScale: TimelineScale) => {
+  return data.map((item) => {
+    const aggregatedChunks = item.chunks.reduce(
+      (currentClusteredEvents: TimebarChartChunk[], currentEvent: TimebarChartChunk) => {
+        const lastClusteredEvent = currentClusteredEvents[currentClusteredEvents.length - 1]
+        const lastType = lastClusteredEvent ? lastClusteredEvent.type : null
+        const lastEnd =
+          lastClusteredEvent && lastClusteredEvent.end
+            ? lastClusteredEvent.end
+            : Number.NEGATIVE_INFINITY
+        const currentEventStartPx = outerScale(currentEvent.start)
+        const lastEndPx = outerScale(lastEnd)
```

Comment:
This is returning NaN when `lastEnd` is Number.NEGATIVE_INFINITY I think we are safe using `deltaPx > MIN_DISTANCE_PX_TO_CLUSTER` but worth mentioning